### DISCUSSION TITLE
Feedback on edgedb in the context of python

### DISCUSSION THREAD

**Discussion by arogozhnikov** [geldata/gel#5122] [General]
Hi edgedb team,

I've evaluated edgedb for my purposes, and decided in favor of sqlmodel/sqlachemy2.

There are multiple improvements in edgedb over plain postgres that attract me, but there is also a number of show-stoppers as well. I think this kind of feedback should be helpful.

I've compiled my summary in this gist: https://gist.github.com/arogozhnikov/41b6ac59fbe094a574ea016c271a7868

Feel free to comment and correct me if I missed something.

---

## Notes on edgedb

I’ve made a test evaluation of edgedb for the purposes of a small company.

My order of priorities (starting from most important): 

1. correctness of results (data and code)
2. developer convenience
3. minimal maintenance 
4. efficiency/scalability

Setup that I compare to: 
postgresql on AWS aurora (managed with AWS CDK) + `sqlalchemy` + `sqlmodel` + `alembic`

### EdgeDB provides a number of significant improvements over reference setup:

- SDL (schema definition language). I’d prefer it over ORM for schema management
- generation of migrations is provided by database tools. Finally
- select part of EdgeQL is expressive and readable
- (multi)links are more friendly than (double)joins
- computeds look extremely helpful, specially links
- access controls also looks very useful, as protecting access on db level is the best option

### Downsides: 

1. there is no way to migrate down (undo migration). 
   I use it quite often, and don’t want to clutter history with experiments. 
   (migrating down mostly happens in local and dev stages)
   
   Migration rewrite is undocumented. Github [discussion](https://github.com/edgedb/edgedb/discussions/4562) did not help me at all.

2. edgedb website really pushes on composability of edgeql, i.e. you can ‘nest’ one query into another (like in graphQL). 

   While the statement is correct, my understanding of composability includes ability to name some piece of code and refer to it by name. 
   In other words, composability is useful when you can decompose into parts and combine them. 
   SQL has temporary tables, views, common table expression - all far from being great tools, but to some degree they satisfy this requirement; edgeql queries don't.
  
   Another example of composability: 
   ORMs can define some expression, and then pass it to outer scope, where expression can be narrowed down by subselecting for data required in this context.
   I don’t see a similar tool in edgedb/edgeql - queries are isolated and can't be combined.
  
3. Python client. That’s the biggest point, and edgedb loses to sqlalchemy/sqlmodel here.

   Official examples (e.g. for Flask) commonly use inline queries, which 
   - are not statically checked against schema. For contrast, pycharm can read and validate sql queries even in python strings 
   - results do not have helpful type hinting.
   
   Both points are a significant downgrade from sqlalchemy2/sqlmodel.
  
   Another available option is code generation. Queries written in .edgeql are wrapped into python code with proper typing.
   This addresses previous problems: code generation verifies query against the schema, and typing is available.
   Inconveniently, every query should be its own file.
  
   Code generation however does not reduce amount of code you need to write. 
   To insert/update/upsert/select by id, you need to write a query with all parameters typed. 
   I like that parameters are typed, but why can’t I just have it autogenerated from schema?
   This could be just a number of typed functions `create_movie(id: Optional[uuid], title: str, …)`, `update_movie(id: uuid, title: Optional[str]=NoValue, …)`
  
   Or even embrace TypedDicts to map objects to deliver help with typing inputs/outputs without ORMs and classes.
   
   As a result, most common operations require a significant hassle in edgedb.
   
   

### Comments: 

- Some other things (uuid, inheritance, insert.. on conflict...) are available in postgres, so I don't mention them
- Typescript client is far more advanced. 
  Well, if I used typescript, I'd probably compare with prisma.
- Syntax for back links doesn't click in my head
- I did not figure out how I manage edgedb with aurora (without cloudformation). Should be solvable.
- EdgeDB ui provides query editor, schema viewer, data viewer/editor.
  It simpolifies onboarding of new developers as they don't need to install any tools.
  Editor, on the other hand is not as convenient/advanced as e.g. IntelliJ's solutions.
- I've made a pass thru python repos on github that depend on edgedb, and none of them made a significant usage of edgedb. Best [match](https://github.com/reizio/reiz.io) actually ended up creating query builder on top of edgeql, in line with my previous points.

**Comment by elprans** [comment]
Hi @arogozhnikov! Thanks a lot for posting feedback, appreciate it!

I took the liberty to re-post your Gist above as it'll hopefully be more convenient to lead the discussion here and easier to find for others.  Some comments in threads below.

**Comment by elprans** [comment]
> While the statement is correct, my understanding of composability includes ability to name some piece of code and refer to it by name. In other words, composability is useful when you can decompose into parts and combine them. SQL has temporary tables, views, common table expression - all far from being great tools, but to some degree they satisfy this requirement; edgeql queries don't.

I guess this could be a discoverability issue.  EdgeDB does have equivalents for views and CTEs, we call them _expression aliases_, which can be declared both at the [schema level](https://www.edgedb.com/docs/datamodel/aliases#ref-datamodel-aliases): 

```
alias NewUsers := (
  select 
    User
  filter
    (datetime_of_statement() - .registration_date) < <duration>'24 hours'
)
```

and at the [query level](https://www.edgedb.com/docs/edgeql/with):

```
with
  movie_actor_names := {"Robert Downey Jr.", "Chris Evans"},
  actors := (select Person filter .name in movie_actor_names),
  avengers := (insert Movie { title := "Avengers", actors := actors }),
  endgame := (insert Movie { title := "Avengers: Endgame", actors := actors }),
  account := (select Account filter .login = <str>$login)
update
  account
set {
  watch_list += {avengers, endgame}
}
```

(the last example is from my 1.0 release day [talk](https://youtu.be/Z0D24lZttvg?t=699))

**Comment by elprans** [comment]
> Another example of composability: ORMs can define some expression, and then pass it to outer scope, where expression can be narrowed down by subselecting for data required in this context.

This mostly stems from the (current) lack of proper Python query builder (which is fair, because you evaluated in the context of Python).  You can do this with the JavaScript binding:

```
function getMoviesQuery() {
  return e.select(e.Movie, (movie) => ({
    id: true,
    title: true,
  }));
}

function getFilteredMovies() {
  return e.select(getMoviesQuery(), (movie) => ({
    filter: e.op(movie.release_year, ">", 1999)
  });
```

**Comment by elprans** [comment]
> Migration rewrite is undocumented.

It's a 3.0 feature, which has not been released yet.  There is a lot of new migration-related features coming in 3.0, the docs are being written now.

**Comment by elprans** [comment]
> To insert/update/upsert/select by id, you need to write a query with all parameters typed. I like that parameters are typed, but why can’t I just have it autogenerated from schema?

This is a great suggestion!  The current codegen is basically v1.0, there's definitely lots of room for making it better.  Please don't hesitate to open issues in [edgedb-python](https://github.com/edgedb/edgedb-python).
